### ISSUE TITLE
Can't resolve path of method declaration

### ISSUE THREAD

**Issue by joined** [javaparser/javaparser#1552]
I'm using JavaParser and JavaSymbolSolver to extract the method calls from a Java class file.
I'm interested in extracting, for each of the method calls within the class:

1. the line range of the method call
2. the qualified signature of the called method
3. the line range of the called method's declaration
4. the path of the java class containing the called method declaration

So far I had success with 1-3 but not so much with 4.
Here is the code that I'm using:

```java
package nl.tud.mcextractor;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
import com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserMethodDeclaration;
import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

import java.io.File;
import java.util.List;

public class Runner {
    private static final String FILE_PATH = "jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java";

    public static void main(String[] args) throws Exception {
        TypeSolver typeSolver = new CombinedTypeSolver(
                new ReflectionTypeSolver(),
                new JavaParserTypeSolver(new File("jNumerus/src/main/java")));

        CompilationUnit cu = JavaParser.parse(new File(FILE_PATH));

        System.out.format("Analyzing method calls of file %s\n", cu.getStorage().get().getPath());

        List<MethodCallExpr> methodCalls = cu.findAll(MethodCallExpr.class);
        for (MethodCallExpr mc : methodCalls) {
            ResolvedMethodDeclaration correspondingDeclaration = JavaParserFacade.get(typeSolver).solve(mc).getCorrespondingDeclaration();
            if (correspondingDeclaration instanceof JavaParserMethodDeclaration) {
                JavaParserMethodDeclaration declaration = (JavaParserMethodDeclaration) correspondingDeclaration;
                Node wrappedNode = declaration.getWrappedNode();
                MethodDeclaration wrappedDeclaration = (MethodDeclaration) wrappedNode;
                System.out.format("L%dC%d,L%dC%d -> %s, (L%dC%d,L%dC%d)\n",
                        mc.getBegin().get().line,
                        mc.getBegin().get().column,
                        mc.getEnd().get().line,
                        mc.getEnd().get().column,
                        correspondingDeclaration.getQualifiedSignature(),
                        wrappedDeclaration.getBegin().get().line,
                        wrappedDeclaration.getBegin().get().column,
                        wrappedDeclaration.getEnd().get().line,
                        wrappedDeclaration.getEnd().get().column);
                Node temp = wrappedNode.findRootNode();
                if (temp instanceof CompilationUnit) {
                    CompilationUnit compilationUnit = (CompilationUnit) temp;
                    if (compilationUnit.getStorage().isPresent()) {
                        System.out.format("Path: %s\n", compilationUnit.getStorage().get().getPath());
                    } else {
                        System.out.println("Couldn't resolve path");
                    }
                }
            }
        }
    }
}
```
the sample project that I'm trying to analyze is this one https://github.com/TheMatjaz/jNumerus.
This is the output:
```
Analyzing method calls of file mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java

L98C9,L98C33 -> it.matjaz.jnumerus.RomanInteger.setValueAndNumeral(int), (L122C5,L125C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L110C9,L110C35 -> it.matjaz.jnumerus.RomanInteger.setNumeralAndValue(it.matjaz.jnumerus.RomanNumeral), (L136C5,L139C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L123C24,L123C61 -> it.matjaz.jnumerus.RomanConverter.integerToRomanNumeral(int), (L186C5,L194C5)
Couldn't resolve path
L137C22,L137C61 -> it.matjaz.jnumerus.RomanConverter.romanNumeralToInteger(it.matjaz.jnumerus.RomanNumeral), (L122C5,L124C5)
Couldn't resolve path
L170C9,L170C33 -> it.matjaz.jnumerus.RomanInteger.setValueAndNumeral(int), (L122C5,L125C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L182C9,L182C35 -> it.matjaz.jnumerus.RomanInteger.setNumeralAndValue(it.matjaz.jnumerus.RomanNumeral), (L136C5,L139C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L223C27,L223C42 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L226C16,L226C54 -> it.matjaz.jnumerus.RomanNumeral.equals(java.lang.Object), (L328C5,L338C5)
Couldn't resolve path
L226C36,L226C53 -> it.matjaz.jnumerus.RomanInteger.getNumeral(), (L155C5,L157C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L239C30,L239C47 -> it.matjaz.jnumerus.RomanNumeral.toString(), (L352C5,L355C5)
Couldn't resolve path
L267C16,L267C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L280C16,L280C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L293C16,L293C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L306C16,L306C25 -> it.matjaz.jnumerus.RomanInteger.getValue(), (L146C5,L148C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L326C13,L326C27 -> it.matjaz.jnumerus.RomanInteger.intValue(), (L265C5,L268C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
L328C20,L328C34 -> it.matjaz.jnumerus.RomanInteger.intValue(), (L265C5,L268C5)
Path: mcextractor/jNumerus/src/main/java/it/matjaz/jnumerus/RomanInteger.java
```
Basically I can extract the path only when the called method is declared in the class that I'm analyzing itself. Any idea on how to extract it also for methods declared in other parts of the codebase?

**Comment by ftomassetti** [comment]
Ciao Lorenzo,
the fact is that JSS can resolve references to methods that come from different sources. When they come from a Java file they are effectively instances of `JavaParserMethodDeclaration`. However they could come from a JAR or obtained through reflection. In those cases there is no Java file to refer to. In your case you are just using a Reflection and a JavaParser typesolver so your references will be either JavaParserMethodDeclaration or ReflectionMethodDeclaration. In the latter case there is no path associated to. If you were also using JARs you could find the path of those JARs but obviously not of the java file because there would be no Java file involved.

**Comment by joined** [comment]
> Ciao Lorenzo,
the fact is that JSS can resolve references to methods that come from different sources. When they come from a Java file they are effectively instances of JavaParserMethodDeclaration. However they could come from a JAR or obtained through reflection. In those cases there is no Java file to refer to. In your case you are just using a Reflection and a JavaParser typesolver so your references will be either JavaParserMethodDeclaration or ReflectionMethodDeclaration. In the latter case there is no path associated to. If you were also using JARs you could find the path of those JARs but obviously not of the java file because there would be no Java file involved.

Thanks for the quick answer @ftomassetti.
I see what you mean but if you look at the code that I've posted, I'm not trying to resolve the path of "reflection" method declarations.
The "couldn't get path" message is printed only when `getCorrespondingDeclaration()` returns a `JavaParserMethodDeclaration`, but then `compilationUnit.getStorage().isPresent()` is false.
The methods that I'm trying to solve the path of are defined in Java classes contained in the same directory of the class that I'm analyzing. That's why I'm confused.

**Comment by ftomassetti** [comment]
ok, sorry about the confusion then. I think @matozoid knows better about the storage feature. It does not seem to be an issue related to symbol solving. Removing the flag

**Comment by matozoid** [comment]
It's related to both :-) PR is open for review.

Note that JavaParserFacade is the old fashioned API, do check out https://github.com/javaparser/javasymbolsolver-maven-sample/blob/master/src/main/java/com/yourorganization/maven_sample/MyAnalysis.java
